java/src/main/native/src/CompiledExpression.cpp

```diff
@@ -0,0 +1,427 @@
+  add_column_ref(std::unique_ptr<cudf::ast::column_reference> ref_ptr) {
+    column_refs.push_back(std::move(ref_ptr));
+    return *column_refs.back();
+  }
+
+  cudf::ast::expression &add_expression(std::unique_ptr<cudf::ast::expression> expr_ptr) {
+    expressions.push_back(std::move(expr_ptr));
+    return *expressions.back();
+  }
+
+  /** Return the expression node at the top of the tree */
+  cudf::ast::expression &get_top_expression() const { return *expressions.back(); }
+};
+
+} // namespace ast
+} // namespace jni
+} // namespace cudf
+
+namespace {
+
+/** Utility class to read data from the serialized AST buffer generated from Java */
+class jni_serialized_ast {
+  jbyte const *data_ptr;      // pointer to the current entity to deserialize
+  jbyte const *const end_ptr; // pointer to the byte immediately after the AST serialized data
+
+  /** Throws an error if there is insufficient space left to read the specified number of bytes */
+  void check_for_eof(std::size_t num_bytes_to_read) {
+    if (data_ptr + num_bytes_to_read > end_ptr) {
+      throw std::runtime_error("Unexpected end of serialized data");
+    }
+  }
+
+public:
+  jni_serialized_ast(cudf::jni::native_jbyteArray &jni_data)
+      : data_ptr(jni_data.begin()), end_ptr(jni_data.end()) {}
+
+  /** Returns true if there is no data remaining to be read */
+  bool at_eof() { return data_ptr == end_ptr; }
+
+  /** Read a byte from the serialized AST data buffer */
+  jbyte read_byte() {
+    check_for_eof(sizeof(jbyte));
+    return *data_ptr++;
+  }
+
+  /** Read a multi-byte value from the serialized AST data buffer */
+  template <typename T> T read() {
+    check_for_eof(sizeof(T));
+    // use memcpy since data may be misaligned
+    T result;
+    memcpy(&result, data_ptr, sizeof(T));
+    data_ptr += sizeof(T);
+    return result;
+  }
+
+  /** Decode a libcudf data type from the serialized AST data buffer */
+  cudf::data_type read_cudf_type() {
+    auto const dtype_id = static_cast<cudf::type_id>(read_byte());
+    switch (dtype_id) {
+      case cudf::type_id::INT8:
+      case cudf::type_id::INT16:
+      case cudf::type_id::INT32:
+      case cudf::type_id::INT64:
+      case cudf::type_id::UINT8:
+      case cudf::type_id::UINT16:
+      case cudf::type_id::UINT32:
+      case cudf::type_id::UINT64:
+      case cudf::type_id::FLOAT32:
+      case cudf::type_id::FLOAT64:
+      case cudf::type_id::BOOL8:
+      case cudf::type_id::TIMESTAMP_DAYS:
+      case cudf::type_id::TIMESTAMP_SECONDS:
+      case cudf::type_id::TIMESTAMP_MILLISECONDS:
+      case cudf::type_id::TIMESTAMP_MICROSECONDS:
+      case cudf::type_id::TIMESTAMP_NANOSECONDS:
+      case cudf::type_id::DURATION_DAYS:
+      case cudf::type_id::DURATION_SECONDS:
+      case cudf::type_id::DURATION_MILLISECONDS:
+      case cudf::type_id::DURATION_MICROSECONDS:
+      case cudf::type_id::DURATION_NANOSECONDS:
+      case cudf::type_id::STRING: return cudf::data_type(dtype_id);
```

Comment:
nit: I am not a big fan of the formatting because of all of the fall through. Especially with this being on the same line as the case for STRING. I missed it on a quick first glance at what was happening.

Reply:
Also a question. We don't support strings anywhere else in the AST code. Why do we support it here? If it is just future proof then I guess it is okay, but it through me off a bit and I had to go back and check that this was actually the case.

Reply:
> nit: I am not a big fan of the formatting because of all of the fall through. 

Agreed.  This is a result of clang formatting which is now strictly enforced.  Looks like adding a new scope block will convince it to not make it a one-liner on the same line as the last `case` statement.

> We don't support strings anywhere else in the AST code. Why do we support it here?

It's for future-proofing since it was easy to handle in those cases.  Same with decimal support which AST doesn't yet handle.  I put in code to support those cases when it was easy to do so for the serialization/deserialization code to make it a bit smoother when adding support for those types in the near future.  Happy to remove that code if you'd rather it not be half-there like it is today.